File: samples/cpp/text_generation/prompt_lookup_decoding_lm.cpp
// Copyright (C) 2023-2025 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

#include <openvino/openvino.hpp>

#include "openvino/genai/llm_pipeline.hpp"

int main(int argc, char* argv[]) try {
    if (3 != argc) {
        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR> '<PROMPT>'");
    }

    ov::genai::GenerationConfig config;
    config.max_new_tokens = 100;
    // Define candidates number for candidate generation
    config.num_assistant_tokens = 5;
    // Define max_ngram_size
    config.max_ngram_size = 3;

    std::string model_path = argv[1];
    std::string prompt = argv[2];
    
    std::string device = "CPU";

    ov::genai::LLMPipeline pipe(
        model_path,
        device,
        ov::genai::prompt_lookup(true));

    auto streamer = [](std::string subword) {
        std::cout << subword << std::flush;
        return ov::genai::StreamingStatus::RUNNING;
    };

    // Since the streamer is set, the results will
    // be printed each time a new token is generated.
    pipe.generate(prompt, config, streamer);
    std::cout << std::endl;
} catch (const std::exception& error) {
    try {
        std::cerr << error.what() << '\n';
    } catch (const std::ios_base::failure&) {}
    return EXIT_FAILURE;
} catch (...) {
    try {
        std::cerr << "Non-exception object thrown\n";
    } catch (const std::ios_base::failure&) {}
    return EXIT_FAILURE;
}

File: samples/python/text_generation/prompt_lookup_decoding_lm.py
#!/usr/bin/env python3
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import argparse
import openvino_genai

def streamer(subword):
    print(subword, end='', flush=True)
    # Return flag corresponds whether generation should be stopped. 
    return openvino_genai.StreamingStatus.RUNNING

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('model_dir')
    parser.add_argument('prompt')
    args = parser.parse_args()

    device = 'CPU'

    pipe = openvino_genai.LLMPipeline(args.model_dir, device, prompt_lookup=True)
    
    config = openvino_genai.GenerationConfig()
    config.max_new_tokens = 100
    # add parameter to enable prompt lookup decoding to generate `num_assistant_tokens` candidates per iteration
    config.num_assistant_tokens = 5
    # Define max_ngram_size
    config.max_ngram_size = 3

    # Since the streamer is set, the results will be printed 
    # every time a new token is generated and put into the streamer queue.
    pipe.generate(args.prompt, config, streamer)
    print()

if '__main__' == __name__:
    main()
